internal/indexer/module_calls.go

```diff
@@ -11,104 +11,111 @@ import (
+	if err != nil {
+		return jobIds, err
+	}
 
-		moduleCalls, err := idx.modStore.ModuleCalls(modHandle.Path())
+	var errs *multierror.Error
+
+	idx.logger.Printf("indexing installed module calls: %d", len(moduleCalls.Installed))
+	for _, mc := range moduleCalls.Installed {
+		fi, err := os.Stat(mc.Path)
+		if err != nil || !fi.IsDir() {
+			multierror.Append(errs, err)
+			continue
+		}
+		err = idx.modStore.Add(mc.Path)
 		if err != nil {
-			return jobIds, err
+			multierror.Append(errs, err)
+			continue
 		}
 
-		var errs *multierror.Error
+		mcHandle := document.DirHandleFromPath(mc.Path)
+		// copy path for queued jobs below
+		mcPath := mc.Path
 
-		for _, mc := range moduleCalls.Installed {
-			fi, err := os.Stat(mc.Path)
-			if err != nil || !fi.IsDir() {
-				multierror.Append(errs, err)
-				continue
-			}
-			err = idx.modStore.Add(mc.Path)
-			if err != nil {
-				multierror.Append(errs, err)
-				continue
-			}
+		refCollectionDeps := make(job.IDs, 0)
 
-			mcHandle := document.DirHandleFromPath(mc.Path)
-			// copy path for queued jobs below
-			mcPath := mc.Path
+		parseId, err := idx.jobStore.EnqueueJob(job.Job{
+			Dir: mcHandle,
+			Func: func(ctx context.Context) error {
+				return module.ParseModuleConfiguration(idx.fs, idx.modStore, mcPath)
+			},
+			Type: op.OpTypeParseModuleConfiguration.String(),
+		})
+		if err != nil {
+			multierror.Append(errs, err)
+		} else {
+			jobIds = append(jobIds, parseId)
+			refCollectionDeps = append(refCollectionDeps, parseId)
+		}
 
-			id, err := idx.jobStore.EnqueueJob(job.Job{
-				Dir: mcHandle,
+		var metaId job.ID
+		if parseId != "" {
+			metaId, err = idx.jobStore.EnqueueJob(job.Job{
+				Dir:  mcHandle,
+				Type: op.OpTypeLoadModuleMetadata.String(),
 				Func: func(ctx context.Context) error {
-					return module.ParseModuleConfiguration(idx.fs, idx.modStore, mcPath)
-				},
-				Type: op.OpTypeParseModuleConfiguration.String(),
-				Defer: func(ctx context.Context, jobErr error) (job.IDs, error) {
-					ids := make(job.IDs, 0)
-
-					id, err := idx.jobStore.EnqueueJob(job.Job{
-						Dir:  mcHandle,
-						Type: op.OpTypeLoadModuleMetadata.String(),
-						Func: func(ctx context.Context) error {
-							return module.LoadModuleMetadata(idx.modStore, mcPath)
-						},
-						Defer: func(ctx context.Context, jobErr error) (job.IDs, error) {
-							return idx.collectReferences(ctx, mcHandle)
-						},
-					})
-					if err != nil {
-						return ids, err
-					} else {
-						ids = append(ids, id)
-					}
-
-					return ids, nil
+					return module.LoadModuleMetadata(idx.modStore, mcPath)
 				},
+				DependsOn: job.IDs{parseId},
 			})
 			if err != nil {
 				multierror.Append(errs, err)
-				continue
+			} else {
+				jobIds = append(jobIds, metaId)
+				refCollectionDeps = append(refCollectionDeps, metaId)
 			}
-			jobIds = append(jobIds, id)
+		}
 
-			id, err = idx.jobStore.EnqueueJob(job.Job{
+		if parseId != "" {
```

Comment:
Do we need module metadata before collecting references? Should we check for `&& metaId != ""` here?

Reply:
It can certainly enrich the references (e.g. because it's used to inform module inputs/outputs). The parsing is however essential and we'd always produce zero references without it. Bear in mind the meaning of this check isn't to draw the dependency, but to allow the reference collection to run. If it does run and the metadata job was successfully scheduled, then it will depend on it too.